### ISSUE TITLE
Is it possible to start a container that use sub-gid created from root user?

### ISSUE THREAD

**Issue by JunoJunho** [lxc/lxc#1622]
# Required information

 * Distribution: Ubuntu 
 * Distribution version: 16.04 LTS
 * The output of
   * `lxc-start --version`  2`.0.7`
   * `lxc-checkconfig` 
```
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-4.4.0-79-generic
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled
Multiple /dev/pts instances: enabled

--- Control groups ---
Cgroup: enabled
Cgroup clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled
Macvlan: enabled
Vlan: enabled
Bridges: enabled
Advanced netfilter: enabled
CONFIG_NF_NAT_IPV4: enabled
CONFIG_NF_NAT_IPV6: enabled
CONFIG_IP_NF_TARGET_MASQUERADE: enabled
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled
FUSE (for use with lxcfs): enabled

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: enabled
```
   * `uname -a` `Linux junho-ubuntu 4.4.0-79-generic #100-Ubuntu SMP Wed May 17 19:58:14 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
   * `cat /proc/self/cgroup`
```
11:memory:/user.slice/user-0.slice/session-14.scope
10:blkio:/user.slice/user-0.slice/session-14.scope
9:freezer:/user/root/0
8:cpu,cpuacct:/user.slice/user-0.slice/session-14.scope
7:hugetlb:/
6:pids:/user.slice/user-0.slice/session-14.scope
5:perf_event:/
4:devices:/user.slice/user-0.slice/session-14.scope
3:net_cls,net_prio:/
2:cpuset:/
1:name=systemd:/user.slice/user-0.slice/session-14.scope
```
   * `cat /proc/1/mounts`
```
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=4006036k,nr_inodes=1001509,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=805268k,mode=755 0 0
/dev/sda2 / ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs rw,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event,release_agent=/run/cgmanager/agents/cgm-release-agent.perf_event 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids,release_agent=/run/cgmanager/agents/cgm-release-agent.pids 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb,release_agent=/run/cgmanager/agents/cgm-release-agent.hugetlb 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=23,pgrp=1,timeout=0,minproto=5,maxproto=5,direct 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime 0 0
sunrpc /run/rpc_pipefs rpc_pipefs rw,relatime 0 0
nfsd /proc/fs/nfsd nfsd rw,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
/dev/sda1 /boot/efi vfat rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
/dev/sdb2 /mountDisk/ext4 ext4 rw,relatime,quota,usrquota,data=ordered 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
cgmfs /run/cgmanager/fs tmpfs rw,relatime,size=100k,mode=755 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=805268k,mode=700,uid=1000,gid=1000 0 0
gvfsd-fuse /run/user/1000/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
tmpfs /run/user/0 tmpfs rw,nosuid,nodev,relatime,size=805268k,mode=700 0 0
/dev/sda2 /var/lib/docker/aufs ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
none /var/lib/docker/aufs/mnt/598a4d87a9f6f5c0ee67f93d31abd797bbe50bcdd3dcc80d8e375e0981c5fe4e aufs rw,relatime,si=b540985d92401de6,dio,dirperm1 0 0
shm /var/lib/docker/containers/610a9e4146ebec5794d48970d3b46078966b0cefa90fb806a20191e80374dc05/shm tmpfs rw,nosuid,nodev,noexec,relatime,size=65536k 0 0
nsfs /run/docker/netns/f16eca51951d nsfs rw 0 0
none /root/container_root/test1 aufs rw,relatime,si=b540985d592f1de6 0 0
```

# Issue description

Hi, I had tried to executing a privileged container using sub-gid mapped configuration.
I created a root privileged container (which means that `lxc-create` by root user (uid:0, gid:0).

After creation, I modified configuration file adding id_map (`lxc.id_map = g 0 10000 65536') to execute new gid (Also set in the /etc/subgid file). 
I think we can still execute the container because we had a root uid (0). However, I typed `lxc-start`, then it cannot start. 

When I tried to start deleted id mapping information, it works well.

# Information to attach

 - [ ] container log (The <log> file from running `lxc-start -n <c> -l <log> -o DEBUG`)
```
      lxc-start 20170613005019.941 INFO     lxc_start_ui - tools/lxc_start.c:main:275 - using rcfile /var/lib/lxc/test1/config
      lxc-start 20170613005019.941 WARN     lxc_confile - confile.c:config_pivotdir:1910 - lxc.pivotdir is ignored.  It will soon become an error.
      lxc-start 20170613005019.941 INFO     lxc_confile - confile.c:config_idmap:1531 - read uid map: type g nsid 0 hostid 10000 range 65536
      lxc-start 20170613005019.943 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:309 - Going to wait for pid 2252.
      lxc-start 20170613005019.944 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:328 - Trying to sync with child process.
      lxc-start 20170613005019.944 INFO     lxc_start - start.c:lxc_check_inherited:235 - Closed inherited fd: 3.
      lxc-start 20170613005019.944 INFO     lxc_start - start.c:lxc_check_inherited:235 - Closed inherited fd: 5.
      lxc-start 20170613005019.944 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:366 - Using pipe file descriptor 6 for monitord.
      lxc-start 20170613005019.948 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:343 - Sucessfully synced with child process.
      lxc-start 20170613005019.949 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:312 - Finished waiting on pid 2252.
      lxc-start 20170613005019.949 INFO     lxc_container - lxccontainer.c:do_lxcapi_start:804 - Attempting to set proc title to [lxc monitor] /var/lib/lxc test1
      lxc-start 20170613005019.949 INFO     lxc_start - start.c:lxc_check_inherited:235 - Closed inherited fd: 3.
      lxc-start 20170613005019.949 INFO     lxc_lsm - lsm/lsm.c:lsm_init:48 - LSM security driver AppArmor
      lxc-start 20170613005019.949 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .reject_force_umount  # comment this to allow umount -f;  not recommended.
      lxc-start 20170613005019.949 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for reject_force_umount action 0.
      lxc-start 20170613005019.949 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:251 - Setting Seccomp rule to reject force umounts.
      lxc-start 20170613005019.949 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for reject_force_umount action 0.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:251 - Setting Seccomp rule to reject force umounts.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .[all].
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .kexec_load errno 1.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for kexec_load action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for kexec_load action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .open_by_handle_at errno 1.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for open_by_handle_at action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for open_by_handle_at action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .init_module errno 1.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for init_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for init_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .finit_module errno 1.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for finit_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for finit_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:402 - processing: .delete_module errno 1.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:567 - Adding native rule for delete_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:570 - Adding compat rule for delete_module action 327681.
      lxc-start 20170613005019.950 INFO     lxc_seccomp - seccomp.c:parse_config_v2:580 - Merging in the compat Seccomp ctx into the main one.
      lxc-start 20170613005019.950 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:309 - Going to wait for pid 2257.
      lxc-start 20170613005019.950 DEBUG    lxc_start - start.c:setup_signal_fd:273 - Set SIGCHLD handler with file descriptor: 5.
      lxc-start 20170613005019.950 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:328 - Trying to sync with child process.
      lxc-start 20170613005019.950 INFO     lxc_start - start.c:lxc_check_inherited:235 - Closed inherited fd: 3.
      lxc-start 20170613005019.950 INFO     lxc_start - start.c:lxc_check_inherited:235 - Closed inherited fd: 5.
      lxc-start 20170613005019.950 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:366 - Using pipe file descriptor 6 for monitord.
      lxc-start 20170613005019.950 DEBUG    console - console.c:lxc_console_peer_default:468 - no console peer
      lxc-start 20170613005019.950 ERROR    lxc_conf - conf.c:chown_mapped_root:3478 - No mapping for container root
      lxc-start 20170613005019.950 ERROR    lxc_conf - conf.c:ttys_shift_ids:3613 - Failed to chown /dev/pts/22
      lxc-start 20170613005019.950 ERROR    lxc_start - start.c:lxc_init:471 - Failed to shift tty into container.
      lxc-start 20170613005019.950 ERROR    lxc_start - start.c:__lxc_start:1302 - Failed to initialize container "test1".
      lxc-start 20170613005019.954 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:343 - Sucessfully synced with child process.
      lxc-start 20170613005019.955 DEBUG    lxc_monitor - monitor.c:lxc_monitord_spawn:312 - Finished waiting on pid 2257.
      lxc-start 20170613005019.955 INFO     lxc_monitor - monitor.c:lxc_monitor_sock_name:185 - Using monitor socket name "lxc/ad055575fe28ddd5//var/lib/lxc".
      lxc-start 20170613005024.955 ERROR    lxc_start_ui - tools/lxc_start.c:main:366 - The container failed to start.
      lxc-start 20170613005024.955 ERROR    lxc_start_ui - tools/lxc_start.c:main:368 - To get more details, run the container in foreground mode.
      lxc-start 20170613005024.955 ERROR    lxc_start_ui - tools/lxc_start.c:main:370 - Additional information can be obtained by setting the --logfile and --logpriority options.
```
 - [ ] the containers configuration file
```
# Common configuration
lxc.include = /usr/share/lxc/config/common.conf
lxc.aa_allow_incomplete = 1
lxc.rootfs = /root/container_root/test1
lxc.rootfs.backend = dir
lxc.init_cmd = /sbin/init
lxc.utsname = test1
lxc.arch = amd64
lxc.cgroup.cpu.shares = 4096
lxc.cgroup.memory.limit_in_bytes = 1024M
lxc.id_map = g 0 10000 65536
#lxc.id_map = u 0 0 65536

lxc.network.type = veth
lxc.network.flags = up
lxc.network.link = bridge_0
lxc.network.name = eno0
lxc.network.veth.pair = vetheno0
lxc.network.hwaddr = aa:bb:cc:dd:ee:11
lxc.network.ipv4 = 192.168.100.101/24
lxc.network.type = veth
lxc.network.flags = up
lxc.network.link = bridge_1
lxc.network.name = eno1
lxc.network.veth.pair = vetheno1
lxc.network.hwaddr = aa:bb:cc:dd:ee:22
lxc.network.ipv4 = 192.168.200.202/24
lxc.network.ipv4.gateway = 192.168.200.2
```

**Comment by brauner** [comment]
So a few notes at first. :)
Let's view a container consisting of privilege rings. 
1. First privilege ring:
There's a first privilege ring which is where the container setup operations take place. That can but does not have to include creating the container, mapping user ids and group ids, creating cgroups, spawning the init binary etc. This can either be done unprivileged, e.g. when your user id is not zero or privileged when your user id is zero (aka when you're root.)
2. Second privilege ring:
 Then there's a second privilege ring which is the running container itself after it has spawned the requested init binary. The container itself can also be either running privileged or unprivileged. When you don't map user ids then the container is running privileged. When you map user ids the container is (barring corner cases where some ids are not mapped or are the identical mapping) the container will run unprivileged.

What people usually refer to when they talk about unprivileged containers is the second ring, i.e. the container is running with a user id and group id mapping. Who actually created the container does not matter in this viewpoint.

Now, to your actual problem. You created a privileged container but did not shift the user ids and group ids of the root filesystem that the container is run in. Then you specified a user id and group id mapping. However, because of not mapping the container's rootfs as well the container does not hold privilege over any of the files in the root filesystem which means it cannot start the init binary. What you actually want to do is to create the container as a root user with a mapping already defined. That is in your

    /etc/lxc/default.conf

you want the same entries that you tried to insert afterwards into your container's configuration file. Please also note, that in order to create an unprivileged container you need to use the download template `-t download` since necessary mknod operations and so on are not permitted when mapping user ids and group ids.

**Comment by JunoJunho** [comment]
What a wonderful note I ever had! :)
Thanks for @brauner 

Unfortunately, I still have a few question about this problem.

1. In unprivileged container, do I have to map both uid and gid? I mean, can I map just on of them? When I see the code in lxc-create, it checks whether both uid and gid have a mapping relationship.
2. You said privilege ring. In my understanding, first ring is about container creation, and second ring is about container execution. So, even though I created lxc-container using root:root (privileged), can I execute those init binary and mknod operation if I change their ownership to mapped id?
For example, if I create a container using root:root, then I use `chown` command to change rootfs's ownership to root:100000 (gid is changed to sub-gid). Then, executing lxc-container using config file in `/var/lib/lxc/container-name/config` with `lxc.id_map = g 0 100000 65536` and `/etc/subgid` with `root:100000`.
I think this has no problem because I changed the ownership of rootfs to subgid or subuid even if I create the container using root account in first privilege ring.
3. Now, an unprivileged container can be created using only template? In other words, we cannot create an unprivileged container without container template provided by lxc?
